MMCore/MMCore.cpp

```diff
@@ -8032,6 +8042,777 @@ std::vector<std::string> CMMCore::getLoadedPeripheralDevices(const char* hubLabe
+   if (storage)
+   {
+      mm::DeviceModuleLockGuard guard(storage);
+      std::string meta;
+		std::vector<int> coords(coordinates.begin(), coordinates.end());
+      int ret = storage->GetImageMeta(handle, coords, meta);
+      if (ret != DEVICE_OK)
+      {
+         logError(getDeviceName(storage).c_str(), getDeviceErrorText(ret, storage).c_str());
+         throw CMMError(getDeviceErrorText(ret, storage).c_str(), MMERR_StorageMetadataNotAvailable);
+      }
+      return meta;
+   }
+   throw CMMError(getCoreErrorText(MMERR_StorageNotAvailable).c_str(), MMERR_StorageNotAvailable);
+}
+
+/**
+ * Sets custom metadata for a given dataset.
+ * Custom metadata is re-writable, i.e. it can be called at any time,
+ * during acquisition or after.
+ *
+ * @param handle   Dataset handle to set metadata for
+ * @param key      Metadata key/name identifier
+ * @param meta     Metadata string to be stored
+ *
+ * @throws MMCoreException If the specified device handle is invalid or key not found
+ */
+void CMMCore::setDatasetCustomMeta(const char* handle, const char* key, const char* meta)
+{
+   std::shared_ptr<StorageInstance> storage = currentStorage_.lock();
+   if (storage)
+   {
+      mm::DeviceModuleLockGuard guard(storage);
+      std::string metastr(meta);
+      int ret = storage->SetCustomMeta(handle, key, metastr);
+      if (ret != DEVICE_OK)
+      {
+         logError(getDeviceName(storage).c_str(), getDeviceErrorText(ret, storage).c_str());
+         throw CMMError("Error writing custom metadata", MMERR_DEVICE_GENERIC);
+      }
+   }
+   throw CMMError(getCoreErrorText(MMERR_StorageNotAvailable).c_str(), MMERR_StorageNotAvailable);
+}
+
+/**
+ * Retrieves custom metadata for a given dataset.
+ *
+ * @param handle    Dataset handle to get metadata from
+ * @param key      Metadata key/name identifier
+ *
+ * @return         Stored metadata string for the specified key
+ * @throws MMCoreException If the specified device handle is invalid or key not found
+ */
+std::string CMMCore::getDatasetCustomMeta(const char* handle, const char* key)
+{
+   std::shared_ptr<StorageInstance> storage = currentStorage_.lock();
+   if (storage)
+   {
+      mm::DeviceModuleLockGuard guard(storage);
+      std::string meta;
+      int ret = storage->GetCustomMeta(handle, key, meta);
+      if (ret != DEVICE_OK)
+      {
+         logError(getDeviceName(storage).c_str(), getDeviceErrorText(ret, storage).c_str());
+         throw CMMError(getDeviceErrorText(ret, storage).c_str(), MMERR_StorageMetadataNotAvailable);
+      }
+      return meta;
+   }
+   throw CMMError(getCoreErrorText(MMERR_StorageNotAvailable).c_str(), MMERR_StorageNotAvailable);
+}
+
+/**
+ * Returns image pixels at specified coordinates
+ * \param handle - dataset handle
+ * \param coordinates - array of coordinates, one for each dimension
+ * \return - image pixels
+ */
+STORAGEIMGOUT CMMCore::getImageFromDataset(const char* handle, const std::vector<long>& coordinates) throw (CMMError)
```

Comment:
Just pointing out that this API forces MMCoreJ and pymmcore to immediately make a copy of the image buffer, because the buffer can go away at any moment. (Strictly speaking, even that is not fully safe because the copy happens outside of the device module lock.)

It would be nice to add (in the future) a way to share ownership of the image buffer, so that it can be kept alive as (say) a NumPy array without making a copy. But this need not be part of the first iteration, because the focus is presumably on getting camera-to-storage streaming working. We might want to think about buffer lifetime in the device API, though (I already commented there).